File: composeApp/src/commonMain/kotlin/com/jankinwu/fntv/client/data/store/ShortcutSettingsStore.kt
package com.jankinwu.fntv.client.data.store

import androidx.compose.ui.input.key.Key
import androidx.compose.ui.input.key.KeyEvent
import androidx.compose.ui.input.key.isAltPressed
import androidx.compose.ui.input.key.isCtrlPressed
import androidx.compose.ui.input.key.isMetaPressed
import androidx.compose.ui.input.key.isShiftPressed
import androidx.compose.ui.input.key.key
import com.jankinwu.fntv.client.currentPlatform
import com.jankinwu.fntv.client.isMacOS
import com.russhwolf.settings.Settings
import com.russhwolf.settings.set

data class ShortcutKey(
    val keyCode: Long,
    val ctrl: Boolean = false,
    val alt: Boolean = false,
    val shift: Boolean = false,
    val meta: Boolean = false
) {
    fun matches(event: KeyEvent): Boolean {
        return event.key.keyCode == keyCode &&
                event.isCtrlPressed == ctrl &&
                event.isAltPressed == alt &&
                event.isShiftPressed == shift &&
                event.isMetaPressed == meta
    }

    fun format(isMac: Boolean): String {
        val modifierText = if (isMac) {
            buildString {
                if (ctrl) append("⌃")
                if (alt) append("⌥")
                if (shift) append("⇧")
                if (meta) append("⌘")
            }
        } else {
            buildList {
                if (ctrl) add("Ctrl")
                if (alt) add("Alt")
                if (shift) add("Shift")
                if (meta) add("Win")
            }.joinToString(" + ")
        }
        val keyLabel = keyLabelForCode(keyCode, isMac)
        return when {
            modifierText.isBlank() -> keyLabel
            isMac -> modifierText + keyLabel
            else -> "$modifierText + $keyLabel"
        }
    }

    fun toStorageString(): String {
        return listOf(keyCode.toString(), ctrl.toString(), alt.toString(), shift.toString(), meta.toString())
            .joinToString("|")
    }

    companion object {
        fun fromStorageString(raw: String?): ShortcutKey? {
            if (raw.isNullOrBlank()) return null
            val parts = raw.split("|")
            if (parts.size != 5) return null
            val keyCode = parts[0].toLongOrNull() ?: return null
            return ShortcutKey(
                keyCode = keyCode,
                ctrl = parts[1].toBooleanStrictOrNull() ?: false,
                alt = parts[2].toBooleanStrictOrNull() ?: false,
                shift = parts[3].toBooleanStrictOrNull() ?: false,
                meta = parts[4].toBooleanStrictOrNull() ?: false
            )
        }

        fun fromKeyEvent(event: KeyEvent): ShortcutKey? {
            if (isModifierKey(event.key)) return null
            return ShortcutKey(
                keyCode = event.key.keyCode,
                ctrl = event.isCtrlPressed,
                alt = event.isAltPressed,
                shift = event.isShiftPressed,
                meta = event.isMetaPressed
            )
        }

        private fun isModifierKey(key: Key): Boolean {
            return key == Key.ShiftLeft ||
                    key == Key.ShiftRight ||
                    key == Key.CtrlLeft ||
                    key == Key.CtrlRight ||
                    key == Key.AltLeft ||
                    key == Key.AltRight ||
                    key == Key.MetaLeft ||
                    key == Key.MetaRight
        }

        private fun keyLabelForCode(keyCode: Long, isMac: Boolean): String {
            return keyLabelMap(keyCode, isMac) ?: "Key $keyCode"
        }

        private fun keyLabelMap(keyCode: Long, isMac: Boolean): String? {
            return when (keyCode) {
                Key.Enter.keyCode -> if (isMac) "⏎" else "Enter"
                Key.NumPadEnter.keyCode -> if (isMac) "⌤" else "Num Enter"
                Key.Spacebar.keyCode -> if (isMac) "Space" else "Space"
                Key.Tab.keyCode -> if (isMac) "⇥" else "Tab"
                Key.Backspace.keyCode -> if (isMac) "⌫" else "Backspace"
                Key.Insert.keyCode -> "Ins"
                Key.Delete.keyCode -> if (isMac) "⌦" else "Delete"
                Key.Home.keyCode -> if (isMac) "↖" else "Home"
                Key.MoveEnd.keyCode -> if (isMac) "↘" else "End"
                Key.PageUp.keyCode -> if (isMac) "⇞" else "Page Up"
                Key.PageDown.keyCode -> if (isMac) "⇟" else "Page Down"
                Key.PrintScreen.keyCode -> "Print Screen"
                Key.ScrollLock.keyCode -> "Scroll Lock"
                Key.MediaPause.keyCode -> "Pause"
                Key.CapsLock.keyCode -> "Caps Lock"
                Key.NumLock.keyCode -> "Num Lock"
                Key.F2.keyCode -> "F2"
                Key.F3.keyCode -> "F3"
                Key.F4.keyCode -> "F4"
                Key.F5.keyCode -> "F5"
                Key.F6.keyCode -> "F6"
                Key.F7.keyCode -> "F7"
                Key.F8.keyCode -> "F8"
                Key.F9.keyCode -> "F9"
                Key.F10.keyCode -> "F10"
                Key.F11.keyCode -> "F11"
                Key.F12.keyCode -> "F12"
                Key.Escape.keyCode -> "Esc"
                Key.DirectionLeft.keyCode -> "←"
                Key.DirectionRight.keyCode -> "→"
                Key.DirectionUp.keyCode -> "↑"
                Key.DirectionDown.keyCode -> "↓"
                Key.MediaPlayPause.keyCode -> if (isMac) "Play/Pause" else "Play/Pause"
                Key.MediaStepBackward.keyCode -> if (isMac) "Prev" else "Prev"
                Key.MediaStepForward.keyCode -> if (isMac) "Next" else "Next"
                Key.VolumeUp.keyCode -> if (isMac) "Volume Up" else "Volume Up"
                Key.VolumeDown.keyCode -> if (isMac) "Volume Down" else "Volume Down"
                Key.VolumeMute.keyCode -> if (isMac) "Mute" else "Mute"
                Key.MediaStop.keyCode -> if (isMac) "Stop" else "Stop"
                Key.Comma.keyCode -> ","
                Key.Period.keyCode -> "."
                Key.Slash.keyCode -> "/"
                Key.Backslash.keyCode -> "\\"
                Key.Semicolon.keyCode -> ";"
                Key.Apostrophe.keyCode -> "'"
                Key.LeftBracket.keyCode -> "["
                Key.RightBracket.keyCode -> "]"
                Key.Grave.keyCode -> "`"
                Key.Minus.keyCode -> "-"
                Key.Equals.keyCode -> "="
                Key.NumPad0.keyCode -> "Num 0"
                Key.NumPad1.keyCode -> "Num 1"
                Key.NumPad2.keyCode -> "Num 2"
                Key.NumPad3.keyCode -> "Num 3"
                Key.NumPad4.keyCode -> "Num 4"
                Key.NumPad5.keyCode -> "Num 5"
                Key.NumPad6.keyCode -> "Num 6"
                Key.NumPad7.keyCode -> "Num 7"
                Key.NumPad8.keyCode -> "Num 8"
                Key.NumPad9.keyCode -> "Num 9"
                Key.NumPadAdd.keyCode -> "Num +"
                Key.NumPadSubtract.keyCode -> "Num -"
                Key.NumPadMultiply.keyCode -> "Num *"
                Key.NumPadDivide.keyCode -> "Num /"
                Key.NumPadDot.keyCode -> "Num ."
                else -> {
                    when (keyCode) {
                        Key.A.keyCode -> "A"
                        Key.B.keyCode -> "B"
                        Key.C.keyCode -> "C"
                        Key.D.keyCode -> "D"
                        Key.E.keyCode -> "E"
                        Key.F.keyCode -> "F"
                        Key.G.keyCode -> "G"
                        Key.H.keyCode -> "H"
                        Key.I.keyCode -> "I"
                        Key.J.keyCode -> "J"
                        Key.K.keyCode -> "K"
                        Key.L.keyCode -> "L"
                        Key.M.keyCode -> "M"
                        Key.N.keyCode -> "N"
                        Key.O.keyCode -> "O"
                        Key.P.keyCode -> "P"
                        Key.Q.keyCode -> "Q"
                        Key.R.keyCode -> "R"
                        Key.S.keyCode -> "S"
                        Key.T.keyCode -> "T"
                        Key.U.keyCode -> "U"
                        Key.V.keyCode -> "V"
                        Key.W.keyCode -> "W"
                        Key.X.keyCode -> "X"
                        Key.Y.keyCode -> "Y"
                        Key.Z.keyCode -> "Z"
                        Key.Zero.keyCode -> "0"
                        Key.One.keyCode -> "1"
                        Key.Two.keyCode -> "2"
                        Key.Three.keyCode -> "3"
                        Key.Four.keyCode -> "4"
                        Key.Five.keyCode -> "5"
                        Key.Six.keyCode -> "6"
                        Key.Seven.keyCode -> "7"
                        Key.Eight.keyCode -> "8"
                        Key.Nine.keyCode -> "9"
                        else -> null
                    }
                }
            }
        }
    }
}

data class ShortcutBinding(
    val primary: ShortcutKey,
    val secondary: ShortcutKey? = null
) {
    fun matches(event: KeyEvent): Boolean {
        return primary.matches(event) || secondary?.matches(event) == true
    }

    fun format(isMac: Boolean): String {
        val primaryLabel = primary.format(isMac)
        val secondaryLabel = secondary?.format(isMac)
        return if (secondaryLabel.isNullOrBlank()) primaryLabel else "$primaryLabel / $secondaryLabel"
    }
}

enum class ShortcutCategory {
    Search,
    Playback
}

enum class ShortcutActionId {
    FocusSearch,
    TogglePlayPause,
    StopPlayback,
    Mute,
    SeekBackward,
    SeekForward,
    VolumeUp,
    VolumeDown,
    ToggleFullscreen,
    ExitFullscreen,
    SearchNext,
    SearchPrev,
    SearchSelect,
    SearchSwitchTab,
    SearchExit
}

data class ShortcutActionDefinition(
    val id: ShortcutActionId,
    val title: String,
    val category: ShortcutCategory,
    val defaultBinding: ShortcutBinding
)

object ShortcutSettingsStore {
    private val settings = Settings()

    private fun scopedKey(rawKey: String): String {
        val guid = UserInfoMemoryCache.guid
        return if (guid.isNullOrBlank()) rawKey else "$guid::$rawKey"
    }

    val definitions = listOf(
        ShortcutActionDefinition(
            id = ShortcutActionId.FocusSearch,
            title = "聚焦搜索输入框",
            category = ShortcutCategory.Search,
            defaultBinding = ShortcutBinding(
                primary = ShortcutKey(Key.Enter.keyCode),
                secondary = ShortcutKey(Key.NumPadEnter.keyCode)
            )
        ),
        ShortcutActionDefinition(
            id = ShortcutActionId.TogglePlayPause,
            title = "播放/暂停",
            category = ShortcutCategory.Playback,
            defaultBinding = ShortcutBinding(
                primary = ShortcutKey(Key.Spacebar.keyCode),
                secondary = ShortcutKey(Key.MediaPlayPause.keyCode)
            )
        ),
//        ShortcutActionDefinition(
//            id = ShortcutActionId.StopPlayback,
//            title = "停止播放",
//            category = ShortcutCategory.Playback,
//            defaultBinding = ShortcutBinding(
//                primary = ShortcutKey(Key.MediaStop.keyCode)
//            )
//        ),
        ShortcutActionDefinition(
            id = ShortcutActionId.Mute,
            title = "静音/取消静音",
            category = ShortcutCategory.Playback,
            defaultBinding = ShortcutBinding(
                primary = ShortcutKey(Key.M.keyCode),
                secondary = ShortcutKey(Key.VolumeMute.keyCode)
            )
        ),
        ShortcutActionDefinition(
            id = ShortcutActionId.SeekBackward,
            title = "快退 10 秒",
            category = ShortcutCategory.Playback,
            defaultBinding = ShortcutBinding(
                primary = ShortcutKey(Key.DirectionLeft.keyCode),
                secondary = ShortcutKey(Key.MediaStepBackward.keyCode)
            )
        ),
        ShortcutActionDefinition(
            id = ShortcutActionId.SeekForward,
            title = "快进 10 秒",
            category = ShortcutCategory.Playback,
            defaultBinding = ShortcutBinding(
                primary = ShortcutKey(Key.DirectionRight.keyCode),
                secondary = ShortcutKey(Key.MediaStepForward.keyCode)
            )
        ),
        ShortcutActionDefinition(
            id = ShortcutActionId.VolumeUp,
            title = "音量增加",
            category = ShortcutCategory.Playback,
            defaultBinding = ShortcutBinding(
                primary = ShortcutKey(Key.DirectionUp.keyCode),
                secondary = ShortcutKey(Key.VolumeUp.keyCode)
            )
        ),
        ShortcutActionDefinition(
            id = ShortcutActionId.VolumeDown,
            title = "音量减少",
            category = ShortcutCategory.Playback,
            defaultBinding = ShortcutBinding(
                primary = ShortcutKey(Key.DirectionDown.keyCode),
                secondary = ShortcutKey(Key.VolumeDown.keyCode)
            )
        ),
        ShortcutActionDefinition(
            id = ShortcutActionId.ToggleFullscreen,
            title = "切换全屏",
            category = ShortcutCategory.Playback,
            defaultBinding = ShortcutBinding(
                primary = ShortcutKey(Key.F.keyCode)
            )
        ),
        ShortcutActionDefinition(
            id = ShortcutActionId.ExitFullscreen,
            title = "退出全屏",
            category = ShortcutCategory.Playback,
            defaultBinding = ShortcutBinding(
                primary = ShortcutKey(Key.Escape.keyCode)
            )
        ),
        ShortcutActionDefinition(
            id = ShortcutActionId.SearchNext,
            title = "下一个搜索项",
            category = ShortcutCategory.Search,
            defaultBinding = ShortcutBinding(
                primary = ShortcutKey(Key.DirectionDown.keyCode)
            )
        ),
        ShortcutActionDefinition(
            id = ShortcutActionId.SearchPrev,
            title = "上一个搜索项",
            category = ShortcutCategory.Search,
            defaultBinding = ShortcutBinding(
                primary = ShortcutKey(Key.DirectionUp.keyCode)
            )
        ),
        ShortcutActionDefinition(
            id = ShortcutActionId.SearchSelect,
            title = "选中搜索项",
            category = ShortcutCategory.Search,
            defaultBinding = ShortcutBinding(
                primary = ShortcutKey(Key.Enter.keyCode),
                secondary = ShortcutKey(Key.NumPadEnter.keyCode)
            )
        ),
        ShortcutActionDefinition(
            id = ShortcutActionId.SearchSwitchTab,
            title = "切换搜索分类",
            category = ShortcutCategory.Search,
            defaultBinding = ShortcutBinding(
                primary = ShortcutKey(Key.Tab.keyCode)
            )
        ),
        ShortcutActionDefinition(
            id = ShortcutActionId.SearchExit,
            title = "退出搜索",
            category = ShortcutCategory.Search,
            defaultBinding = ShortcutBinding(
                primary = ShortcutKey(Key.Escape.keyCode)
            )
        )
    )

    private val definitionMap = definitions.associateBy { it.id }

    fun getBinding(actionId: ShortcutActionId): ShortcutBinding {
        val definition = definitionMap[actionId] ?: error("Unknown shortcut action: $actionId")
        val primaryRaw = settings.getString(scopedKey("${actionId.name}.primary"), "")
        val secondaryRaw = settings.getString(scopedKey("${actionId.name}.secondary"), "")
        val primary = ShortcutKey.fromStorageString(primaryRaw) ?: definition.defaultBinding.primary
        val secondary = ShortcutKey.fromStorageString(secondaryRaw) ?: definition.defaultBinding.secondary
        return ShortcutBinding(primary = primary, secondary = secondary)
    }

    fun getAllBindings(): Map<ShortcutActionId, ShortcutBinding> {
        return definitions.associate { it.id to getBinding(it.id) }
    }

    fun setBinding(actionId: ShortcutActionId, binding: ShortcutBinding) {
        settings[scopedKey("${actionId.name}.primary")] = binding.primary.toStorageString()
        val secondaryValue = binding.secondary?.toStorageString().orEmpty()
        settings[scopedKey("${actionId.name}.secondary")] = secondaryValue
    }

    fun resetToDefaults() {
        definitions.forEach { definition ->
            setBinding(definition.id, definition.defaultBinding)
        }
    }

    fun matches(event: KeyEvent, actionId: ShortcutActionId): Boolean {
        return getBinding(actionId).matches(event)
    }

    fun shouldSuppressFocusSearchInput(event: KeyEvent): Boolean {
        if (event.isCtrlPressed || event.isAltPressed || event.isMetaPressed) return false
        return isTextInputKey(event.key)
    }

    fun isMacPlatform(): Boolean {
        return currentPlatform().isMacOS()
    }

    private fun isTextInputKey(key: Key): Boolean {
        return when (key) {
            Key.Spacebar,
            Key.A, Key.B, Key.C, Key.D, Key.E, Key.F, Key.G, Key.H, Key.I, Key.J, Key.K, Key.L,
            Key.M, Key.N, Key.O, Key.P, Key.Q, Key.R, Key.S, Key.T, Key.U, Key.V, Key.W, Key.X,
            Key.Y, Key.Z,
            Key.Zero, Key.One, Key.Two, Key.Three, Key.Four, Key.Five, Key.Six, Key.Seven, Key.Eight, Key.Nine -> true
            else -> false
        }
    }
}

File: lib/data/storage/shortcut_settings_store.dart
import 'dart:io' show Platform;

import 'package:flutter/services.dart';
import 'package:shared_preferences/shared_preferences.dart';

import 'preferences_manager.dart';

enum ShortcutCategory { search, playback }

enum ShortcutActionId {
  focusSearch,
  togglePlayPause,
  mute,
  seekBackward,
  seekForward,
  volumeUp,
  volumeDown,
  toggleFullscreen,
  exitFullscreen,
  searchNext,
  searchPrev,
  searchSelect,
  searchSwitchTab,
  searchExit,
}

class ShortcutKeyBinding {
  final int keyId;
  final bool ctrl;
  final bool alt;
  final bool shift;
  final bool meta;

  const ShortcutKeyBinding(
    this.keyId, {
    this.ctrl = false,
    this.alt = false,
    this.shift = false,
    this.meta = false,
  });

  bool matches(KeyEvent event) {
    final keyboard = HardwareKeyboard.instance;
    return event.logicalKey.keyId == keyId &&
        keyboard.isControlPressed == ctrl &&
        keyboard.isAltPressed == alt &&
        keyboard.isShiftPressed == shift &&
        keyboard.isMetaPressed == meta;
  }

  String format({bool? isMac}) {
    final mac = isMac ?? Platform.isMacOS;
    final modifierText = mac
        ? '${ctrl ? '⌃' : ''}${alt ? '⌥' : ''}${shift ? '⇧' : ''}${meta ? '⌘' : ''}'
        : [
            if (ctrl) 'Ctrl',
            if (alt) 'Alt',
            if (shift) 'Shift',
            if (meta) 'Win',
          ].join(' + ');
    final keyLabel = _keyLabelForId(keyId, mac);
    if (modifierText.isEmpty) return keyLabel;
    return mac ? '$modifierText$keyLabel' : '$modifierText + $keyLabel';
  }

  String toStorageString() => '$keyId|$ctrl|$alt|$shift|$meta';

  static ShortcutKeyBinding? fromStorageString(String? raw) {
    if (raw == null || raw.trim().isEmpty) return null;
    final parts = raw.split('|');
    if (parts.length != 5) return null;
    final keyId = int.tryParse(parts[0]);
    if (keyId == null) return null;
    return ShortcutKeyBinding(
      keyId,
      ctrl: bool.tryParse(parts[1]) ?? false,
      alt: bool.tryParse(parts[2]) ?? false,
      shift: bool.tryParse(parts[3]) ?? false,
      meta: bool.tryParse(parts[4]) ?? false,
    );
  }

  static ShortcutKeyBinding? fromKeyEvent(KeyEvent event) {
    if (_isModifierKey(event.logicalKey)) return null;
    final keyboard = HardwareKeyboard.instance;
    return ShortcutKeyBinding(
      event.logicalKey.keyId,
      ctrl: keyboard.isControlPressed,
      alt: keyboard.isAltPressed,
      shift: keyboard.isShiftPressed,
      meta: keyboard.isMetaPressed,
    );
  }

  static bool _isModifierKey(LogicalKeyboardKey key) {
    return key == LogicalKeyboardKey.shiftLeft ||
        key == LogicalKeyboardKey.shiftRight ||
        key == LogicalKeyboardKey.controlLeft ||
        key == LogicalKeyboardKey.controlRight ||
        key == LogicalKeyboardKey.altLeft ||
        key == LogicalKeyboardKey.altRight ||
        key == LogicalKeyboardKey.metaLeft ||
        key == LogicalKeyboardKey.metaRight;
  }

  static String _keyLabelForId(int keyId, bool isMac) {
    final key = LogicalKeyboardKey.findKeyByKeyId(keyId);
    if (key == null) return 'Key $keyId';
    if (key == LogicalKeyboardKey.enter) return isMac ? '⏎' : 'Enter';
    if (key == LogicalKeyboardKey.numpadEnter) return isMac ? '⌤' : 'Num Enter';
    if (key == LogicalKeyboardKey.space) return 'Space';
    if (key == LogicalKeyboardKey.tab) return isMac ? '⇥' : 'Tab';
    if (key == LogicalKeyboardKey.backspace) return isMac ? '⌫' : 'Backspace';
    if (key == LogicalKeyboardKey.delete) return isMac ? '⌦' : 'Delete';
    if (key == LogicalKeyboardKey.home) return isMac ? '↖' : 'Home';
    if (key == LogicalKeyboardKey.end) return isMac ? '↘' : 'End';
    if (key == LogicalKeyboardKey.pageUp) return isMac ? '⇞' : 'Page Up';
    if (key == LogicalKeyboardKey.pageDown) return isMac ? '⇟' : 'Page Down';
    if (key == LogicalKeyboardKey.escape) return 'Esc';
    if (key == LogicalKeyboardKey.arrowLeft) return '←';
    if (key == LogicalKeyboardKey.arrowRight) return '→';
    if (key == LogicalKeyboardKey.arrowUp) return '↑';
    if (key == LogicalKeyboardKey.arrowDown) return '↓';
    if (key == LogicalKeyboardKey.mediaPlayPause) return 'Play/Pause';
    if (key == LogicalKeyboardKey.mediaTrackPrevious) return 'Prev';
    if (key == LogicalKeyboardKey.mediaTrackNext) return 'Next';
    if (key == LogicalKeyboardKey.audioVolumeUp) return 'Volume Up';
    if (key == LogicalKeyboardKey.audioVolumeDown) return 'Volume Down';
    if (key == LogicalKeyboardKey.audioVolumeMute) return 'Mute';

    final label = key.keyLabel;
    if (label.isNotEmpty) return label;
    return key.debugName ?? 'Key $keyId';
  }
}

class ShortcutBinding {
  final ShortcutKeyBinding primary;
  final ShortcutKeyBinding? secondary;

  const ShortcutBinding({required this.primary, this.secondary});

  bool matches(KeyEvent event) {
    return primary.matches(event) || secondary?.matches(event) == true;
  }

  String format({bool? isMac}) {
    final primaryLabel = primary.format(isMac: isMac);
    final secondaryLabel = secondary?.format(isMac: isMac);
    if (secondaryLabel == null || secondaryLabel.isEmpty) return primaryLabel;
    return '$primaryLabel / $secondaryLabel';
  }
}

class ShortcutActionDefinition {
  final ShortcutActionId id;
  final String title;
  final ShortcutCategory category;
  final ShortcutBinding defaultBinding;

  const ShortcutActionDefinition({
    required this.id,
    required this.title,
    required this.category,
    required this.defaultBinding,
  });
}

class ShortcutSettingsStore {
  final SharedPreferences _prefs;
  final String? _userGuid;

  ShortcutSettingsStore(this._prefs, {String? userGuid})
      : _userGuid = PreferencesManager.normalizeGuid(userGuid);

  static final definitions = <ShortcutActionDefinition>[
    ShortcutActionDefinition(
      id: ShortcutActionId.focusSearch,
      title: '聚焦搜索输入框',
      category: ShortcutCategory.search,
      defaultBinding: ShortcutBinding(
        primary: ShortcutKeyBinding(LogicalKeyboardKey.enter.keyId),
        secondary: ShortcutKeyBinding(LogicalKeyboardKey.numpadEnter.keyId),
      ),
    ),
    ShortcutActionDefinition(
      id: ShortcutActionId.togglePlayPause,
      title: '播放/暂停',
      category: ShortcutCategory.playback,
      defaultBinding: ShortcutBinding(
        primary: ShortcutKeyBinding(LogicalKeyboardKey.space.keyId),
        secondary: ShortcutKeyBinding(LogicalKeyboardKey.mediaPlayPause.keyId),
      ),
    ),
    ShortcutActionDefinition(
      id: ShortcutActionId.mute,
      title: '静音/取消静音',
      category: ShortcutCategory.playback,
      defaultBinding: ShortcutBinding(
        primary: ShortcutKeyBinding(LogicalKeyboardKey.keyM.keyId),
        secondary: ShortcutKeyBinding(LogicalKeyboardKey.audioVolumeMute.keyId),
      ),
    ),
    ShortcutActionDefinition(
      id: ShortcutActionId.seekBackward,
      title: '快退 10 秒',
      category: ShortcutCategory.playback,
      defaultBinding: ShortcutBinding(
        primary: ShortcutKeyBinding(LogicalKeyboardKey.arrowLeft.keyId),
        secondary:
            ShortcutKeyBinding(LogicalKeyboardKey.mediaTrackPrevious.keyId),
      ),
    ),
    ShortcutActionDefinition(
      id: ShortcutActionId.seekForward,
      title: '快进 10 秒',
      category: ShortcutCategory.playback,
      defaultBinding: ShortcutBinding(
        primary: ShortcutKeyBinding(LogicalKeyboardKey.arrowRight.keyId),
        secondary: ShortcutKeyBinding(LogicalKeyboardKey.mediaTrackNext.keyId),
      ),
    ),
    ShortcutActionDefinition(
      id: ShortcutActionId.volumeUp,
      title: '音量增加',
      category: ShortcutCategory.playback,
      defaultBinding: ShortcutBinding(
        primary: ShortcutKeyBinding(LogicalKeyboardKey.arrowUp.keyId),
        secondary: ShortcutKeyBinding(LogicalKeyboardKey.audioVolumeUp.keyId),
      ),
    ),
    ShortcutActionDefinition(
      id: ShortcutActionId.volumeDown,
      title: '音量减少',
      category: ShortcutCategory.playback,
      defaultBinding: ShortcutBinding(
        primary: ShortcutKeyBinding(LogicalKeyboardKey.arrowDown.keyId),
        secondary: ShortcutKeyBinding(LogicalKeyboardKey.audioVolumeDown.keyId),
      ),
    ),
    ShortcutActionDefinition(
      id: ShortcutActionId.toggleFullscreen,
      title: '切换全屏',
      category: ShortcutCategory.playback,
      defaultBinding: ShortcutBinding(
        primary: ShortcutKeyBinding(LogicalKeyboardKey.keyF.keyId),
      ),
    ),
    ShortcutActionDefinition(
      id: ShortcutActionId.exitFullscreen,
      title: '退出全屏',
      category: ShortcutCategory.playback,
      defaultBinding: ShortcutBinding(
        primary: ShortcutKeyBinding(LogicalKeyboardKey.escape.keyId),
      ),
    ),
    ShortcutActionDefinition(
      id: ShortcutActionId.searchNext,
      title: '下一个搜索项',
      category: ShortcutCategory.search,
      defaultBinding: ShortcutBinding(
        primary: ShortcutKeyBinding(LogicalKeyboardKey.arrowDown.keyId),
      ),
    ),
    ShortcutActionDefinition(
      id: ShortcutActionId.searchPrev,
      title: '上一个搜索项',
      category: ShortcutCategory.search,
      defaultBinding: ShortcutBinding(
        primary: ShortcutKeyBinding(LogicalKeyboardKey.arrowUp.keyId),
      ),
    ),
    ShortcutActionDefinition(
      id: ShortcutActionId.searchSelect,
      title: '选中搜索项',
      category: ShortcutCategory.search,
      defaultBinding: ShortcutBinding(
        primary: ShortcutKeyBinding(LogicalKeyboardKey.enter.keyId),
        secondary: ShortcutKeyBinding(LogicalKeyboardKey.numpadEnter.keyId),
      ),
    ),
    ShortcutActionDefinition(
      id: ShortcutActionId.searchSwitchTab,
      title: '切换搜索分类',
      category: ShortcutCategory.search,
      defaultBinding: ShortcutBinding(
        primary: ShortcutKeyBinding(LogicalKeyboardKey.tab.keyId),
      ),
    ),
    ShortcutActionDefinition(
      id: ShortcutActionId.searchExit,
      title: '退出搜索',
      category: ShortcutCategory.search,
      defaultBinding: ShortcutBinding(
        primary: ShortcutKeyBinding(LogicalKeyboardKey.escape.keyId),
      ),
    ),
  ];

  static final Map<ShortcutActionId, ShortcutActionDefinition> _definitionMap =
      {
    for (final definition in definitions) definition.id: definition,
  };

  ShortcutBinding getBinding(ShortcutActionId actionId) {
    final definition = _definitionMap[actionId];
    if (definition == null) {
      throw ArgumentError('Unknown shortcut action: $actionId');
    }
    final primary = ShortcutKeyBinding.fromStorageString(
          _readStringScoped(_storageKey(actionId, 'primary')),
        ) ??
        definition.defaultBinding.primary;
    final secondary = ShortcutKeyBinding.fromStorageString(
          _readStringScoped(_storageKey(actionId, 'secondary')),
        ) ??
        definition.defaultBinding.secondary;
    return ShortcutBinding(primary: primary, secondary: secondary);
  }

  Map<ShortcutActionId, ShortcutBinding> getAllBindings() {
    return {
      for (final definition in definitions)
        definition.id: getBinding(definition.id),
    };
  }

  Future<void> setBinding(
    ShortcutActionId actionId,
    ShortcutBinding binding,
  ) async {
    await _writeStringScoped(
      _storageKey(actionId, 'primary'),
      binding.primary.toStorageString(),
    );
    await _writeStringScoped(
      _storageKey(actionId, 'secondary'),
      binding.secondary?.toStorageString() ?? '',
    );
  }

  Future<void> resetToDefaults() async {
    for (final definition in definitions) {
      await setBinding(definition.id, definition.defaultBinding);
    }
  }

  bool matches(KeyEvent event, ShortcutActionId actionId) {
    return getBinding(actionId).matches(event);
  }

  bool shouldSuppressFocusSearchInput(KeyEvent event) {
    final keyboard = HardwareKeyboard.instance;
    if (keyboard.isControlPressed ||
        keyboard.isAltPressed ||
        keyboard.isMetaPressed) {
      return false;
    }
    return _isTextInputKey(event.logicalKey);
  }

  String _storageKey(ShortcutActionId actionId, String slot) {
    return 'shortcut.${actionId.name}.$slot';
  }

  String _scopedKey(String rawKey) {
    final guid = _userGuid;
    if (guid == null) return rawKey;
    return '$guid::$rawKey';
  }

  // 作用域读取：未登录读全局键；登录态只读 <guid>::<key>。
  // 不再做"懒迁移"复制：迁移由 UserSettingsMigrator 统一处理并删除全局值。
  String? _readStringScoped(String rawKey) {
    final guid = _userGuid;
    if (guid == null) {
      return _prefs.getString(rawKey);
    }
    return _prefs.getString('$guid::$rawKey');
  }

  Future<void> _writeStringScoped(String rawKey, String value) {
    return _prefs.setString(_scopedKey(rawKey), value);
  }

  bool _isTextInputKey(LogicalKeyboardKey key) {
    return key == LogicalKeyboardKey.space ||
        key == LogicalKeyboardKey.keyA ||
        key == LogicalKeyboardKey.keyB ||
        key == LogicalKeyboardKey.keyC ||
        key == LogicalKeyboardKey.keyD ||
        key == LogicalKeyboardKey.keyE ||
        key == LogicalKeyboardKey.keyF ||
        key == LogicalKeyboardKey.keyG ||
        key == LogicalKeyboardKey.keyH ||
        key == LogicalKeyboardKey.keyI ||
        key == LogicalKeyboardKey.keyJ ||
        key == LogicalKeyboardKey.keyK ||
        key == LogicalKeyboardKey.keyL ||
        key == LogicalKeyboardKey.keyM ||
        key == LogicalKeyboardKey.keyN ||
        key == LogicalKeyboardKey.keyO ||
        key == LogicalKeyboardKey.keyP ||
        key == LogicalKeyboardKey.keyQ ||
        key == LogicalKeyboardKey.keyR ||
        key == LogicalKeyboardKey.keyS ||
        key == LogicalKeyboardKey.keyT ||
        key == LogicalKeyboardKey.keyU ||
        key == LogicalKeyboardKey.keyV ||
        key == LogicalKeyboardKey.keyW ||
        key == LogicalKeyboardKey.keyX ||
        key == LogicalKeyboardKey.keyY ||
        key == LogicalKeyboardKey.keyZ ||
        key == LogicalKeyboardKey.digit0 ||
        key == LogicalKeyboardKey.digit1 ||
        key == LogicalKeyboardKey.digit2 ||
        key == LogicalKeyboardKey.digit3 ||
        key == LogicalKeyboardKey.digit4 ||
        key == LogicalKeyboardKey.digit5 ||
        key == LogicalKeyboardKey.digit6 ||
        key == LogicalKeyboardKey.digit7 ||
        key == LogicalKeyboardKey.digit8 ||
        key == LogicalKeyboardKey.digit9;
  }
}
